### PR TITLE
Do not copy libstd dynamic library to sysroot

### PR DESCRIPTION
**Pull request by Kobzol**
Since https://github.com/rust-lang/rust/pull/122362, rustc links statically to libstd.[so|dll]. Which means that the libstd.[so|dll] file no longer has to be in the rustc sysroot. However, we are currently still shipping this file, in every new release of Rust, for no reason, it's just wasted bytes.

This PR removes the dynamic library file from the built sysroot.

However, it is not yet performed on Windows, because stage0 incremental tests start failing there (see description of the issue [here](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Failing.20incr.20tests.20on.20Windows.20when.20std.2Edll.20is.20missing/near/474507064)).

This is an extended version of https://github.com/rust-lang/rust/pull/128986.
CC @Zoxc

### CODE REVIEW

File: src/bootstrap/src/bin/rustc.rs [hunk 1]
```diff
@@ -95,7 +95,6 @@ fn main() {
     // When statically linking `std` into `rustc_driver`, remove `-C prefer-dynamic`
     if env::var("RUSTC_LINK_STD_INTO_RUSTC_DRIVER").unwrap() == "1"
         && crate_name == Some("rustc_driver")
-        && stage != "0"
```
Comment (onur-ozkan) [inline src/bootstrap/src/bin/rustc.rs:98]: Why?

Comment (Kobzol) [reply src/bootstrap/src/bin/rustc.rs:98]: This was added back when the beta compiler did not use static linking. Now beta rustc statically links to libstd, so this shouldn't be needed anymore, AFAIK.

Comment (onur-ozkan) [reply src/bootstrap/src/bin/rustc.rs:98]: stage0 can be configured to custom compiler which uses dynamic linking.

Comment (Kobzol) [reply src/bootstrap/src/bin/rustc.rs:98]: Right, but relatively soon most rustc compilers being used should statically link to their libstd. Do we need to support old(er) versions?

Comment (onur-ozkan) [reply src/bootstrap/src/bin/rustc.rs:98]: It's not only about older versions. I just don't want to get a Zulip notification from people building with their custom stage0 compiler and running into this issue. If it's not causing to too much complexity on bootstrap, supporting as many cases as possible is always a good thing in my opinion.

Comment (Kobzol) [reply src/bootstrap/src/bin/rustc.rs:98]: Currently bootstrap produces the statically linked rustc unconditionally, IIRC (on supported platforms), so "custom stage0 compiler" would have to be something where people modified their bootstrap code, at which point anything could happen I suppose.

Without removing this condition, the `stage1/lib` directory would still be required to contain `libstd.so`. That's both surprising, because it does not match what happens on later stages, and it could also conceal potential test failures (or tests succeeding unexpectedly), as is the test that I had to fix in this PR.

As usually, the stage numbering is confusing here - removing the condition alters what happens with the stage1 compiler.

Comment (onur-ozkan) [reply src/bootstrap/src/bin/rustc.rs:98]: > Currently bootstrap produces the statically linked rustc unconditionally

If that's the case yeah this seems reasonable, but what about windows-gnu target?

https://github.com/rust-lang/rust/blob/56e35a5dbb37898433a43133dff0398f46d577b8/src/bootstrap/src/core/builder.rs#L1110-L1114

Comment (Kobzol) [reply src/bootstrap/src/bin/rustc.rs:98]: That will not receive `RUSTC_LINK_STD_INTO_RUSTC_DRIVER`, so it will be still linked dynamically, regardless of the stage :)

File: tests/ui/command/command-current-dir.rs [hunk 2]
```diff
@@ -1,4 +1,5 @@
 //@ run-pass
+//@ no-prefer-dynamic We move the binary around, so do not depend dynamically on libstd
```
Comment (Zoxc) [inline tests/ui/command/command-current-dir.rs:2]: Why is this changed needed? Was it linking to the compiler's std copy before? If so, why?

Comment (Kobzol) [reply tests/ui/command/command-current-dir.rs:2]: The test was buggy. It compiled a Rust binary that dynamically linked to libstd.so. Then it moved the binary around, which should have broken it. However, the binary somehow linked to the compiler's std copy instead (!), so it worked by accident.

File: src/bootstrap/src/bin/rustc.rs (hunk 1 above)
Comment (Zoxc) [reply src/bootstrap/src/bin/rustc.rs:98]: This stage0 check is to ensure support for the `rustc_private` linkage change, which is now present in bootstrap, even on `windows-gnu`.

Comment (onur-ozkan) [reply src/bootstrap/src/bin/rustc.rs:98]: Removal needs to be reverted then. Can you add a comment saying that this check is used for respecting stage0's linkage mode?

Comment (onur-ozkan) [reply src/bootstrap/src/bin/rustc.rs:98]: FWIW linkage mode should be handled with a config option instead of setting that unconditionally IMO, but that's another topic.

File: tests/ui/command/command-current-dir.rs (hunk 2 above)
Comment (lqd) [reply tests/ui/command/command-current-dir.rs:2]: Since UI tests are built with `-Cprefer-dynamic` did they all (modulo `no-prefer-dynamic`) link to the std?

File: src/bootstrap/src/bin/rustc.rs (hunk 1 above)
Comment (lqd) [reply src/bootstrap/src/bin/rustc.rs:98]: > FWIW linkage mode should be handled with a config option instead of setting that unconditionally IMO

I think the 3 of us here would disagree on this ^^

Comment (onur-ozkan) [reply src/bootstrap/src/bin/rustc.rs:98]: > > FWIW linkage mode should be handled with a config option instead of setting that unconditionally IMO
> 
> 
> 
> I think the 3 of us here would disagree on this ^^

What is it so bad about it? You can still keep the current behaviour as a default.

Comment (lqd) [reply src/bootstrap/src/bin/rustc.rs:98]: I consider the previous behavior a hack and am happy to see it disappear. I don't know of any benefits to it.

**Comment by klensy** [comment]
> Since https://github.com/rust-lang/rust/pull/122362, rustc links statically to libstd.[so|dll]

Well, for CI one (x86_64-pc-windows-msvc) - yes, but my local copy for some reason links dynamically (messed up config.toml, probably?)

File: src/bootstrap/src/bin/rustc.rs (hunk 1 above)
Comment (Kobzol) [reply src/bootstrap/src/bin/rustc.rs:98]: > FWIW linkage mode should be handled with a config option instead of setting that unconditionally IMO, but that's another topic.

There's AFAIK absolutely no reason why rustc should link to libstd dynamically. The libstd is tied 1:1 to librustc_driver, it can't really be replaced by a different version. Linking statically makes the compiler faster and reduces the size of the compiler sysroot, which saves us $ :D Sadly, we cannot use it everywhere yet, because of the windows problems, but that's a bug, which should be worked around with an if condition in bootstrap, rather than a config option, IMO.

Comment (onur-ozkan) [reply src/bootstrap/src/bin/rustc.rs:98]: I get that we're not interested in dynamic linking and I never said we were. Adding a config option doesn't mean that we will use it in our CI or default profiles. It’s just about how we design our build system; trying to make it as configurable as possible. Is the reason this isn't configurable because we don't want anyone in the world to use static linking? If that’s the case, fine. But tbh I haven't seen any solid reason that justifies not having a config option so far.

Comment (Zoxc) [reply src/bootstrap/src/bin/rustc.rs:98]: > Removal needs to be reverted then. Can you add a comment saying that this check is used for respecting stage0's linkage mode?

No, it was just there temporarily until the bootstrap compiler was bumped.

Comment (Kobzol) [reply src/bootstrap/src/bin/rustc.rs:98]: > Is the reason this isn't configurable because we don't want anyone in the world to use static linking?

I would in fact argue that the anwer to `Is the reason this isn't configurable because we don't want anyone in the world to use static linking?` is "yes" (well, it should have said "dynamic" instead of "static", but that was a typo :) ). As long as static linking works for your target, I don't think that there is any reason to link to libstd dynamically. The default was swapped a few months ago and I haven't seen anyone complaining so far.

I'm not opposed to making this configurable, but I wouldn't do it proactively. If someone comes with a reasonable use-case why they need it, then sure. It's possible that there are indeed use-cases for this, but I'd like to see them first. We already have a lot of config options and I don't think we necessarily have to add new ones without a specific use-case.

File: tests/ui/command/command-current-dir.rs (hunk 2 above)
Comment (Kobzol) [reply tests/ui/command/command-current-dir.rs:2]: I don't think so, no other tests broke. It was probably just caused by the fact that this binary was moved around, lost its link to the target library and then somehow got dynamically linked to the compiler one instead.

File: src/bootstrap/src/bin/rustc.rs (hunk 1 above)
Comment (onur-ozkan) [reply src/bootstrap/src/bin/rustc.rs:98]: > I'm not opposed to making this configurable, but I wouldn't do it proactively. If someone comes with a reasonable use-case why they need it, then sure. It's possible that there are indeed use-cases for this, but I'd like to see them first. We already have a lot of config options and I don't think we necessarily have to add new ones without a specific use-case.

Yeah this makes more sense.

File: src/bootstrap/src/core/build_steps/compile.rs [hunk 3]
```diff
@@ -1923,8 +1923,24 @@ impl Step for Assemble {
...
         for f in builder.read_dir(&src_libdir) {
             let filename = f.file_name().into_string().unwrap();
-            if (is_dylib(&filename) || is_debug_info(&filename)) && !proc_macros.contains(&filename)
+
+            let is_proc_macro = proc_macros.contains(&filename);
+            let is_dylib_or_debug = is_dylib(&filename) || is_debug_info(&filename);
+
+            // If we link statically to stdlib, do not copy the libstd dynamic library file
+            // Currently, we do not avoid the copy on Windows, as it seems to be causing issues in
+            // post-optimization stage0 tests.
```
[the remark this answers is not in the document]
Comment (Kobzol) [reply src/bootstrap/src/core/build_steps/compile.rs]: Good point. Done that.

**Comment by onur-ozkan** [comment]
You can r=me if you want.

**Comment by Kobzol** [comment]
As usually with these kinds of changes, it's hard to predict all the repercussions that this can make. Let's see if full CI passes, and if yes, if someone starts to complain.

@bors r=onur-ozkan rollup=never

**Comment by madsmtm** [comment]
Since you wrote:

> As usually with these kinds of changes, it's hard to predict all the repercussions that this can have. Let's see if full CI passes, and if yes, if someone starts to complain.

I'll note that this broke testing proc macros with `-Zbuild-std`, see https://github.com/rust-lang/cargo/issues/14735.

That said, this is a minor use-case only affecting nightly users, and Cargo was wrong here before, so I wouldn't argue that this warrants a revert (again, just noting it for future reference).

**Comment by Kobzol** [comment]
Thanks for the heads-up! This is unfortunate, but I also think it's a bug in Cargo (one of many that plague build-std, as it currently stands). We also had some cases where tests were inadvertedly linking to the compiler's libstd instead of the target libstd.
